checkpoint/querier.go

```diff
@@ -113,3 +117,36 @@ func handleQueryCheckpointList(ctx sdk.Context, req abci.RequestQuery, keeper Ke
 	}
 	return bz, nil
 }
+
+func handleQueryNextCheckpoint(ctx sdk.Context, req abci.RequestQuery, keeper Keeper, sk staking.Keeper) ([]byte, sdk.Error) {
+	// get validator set
+	validatorSet := sk.GetValidatorSet(ctx)
+	proposer := validatorSet.GetProposer()
+	ackCount := keeper.GetACKCount(ctx)
+	var start uint64
+	if ackCount != 0 {
+		headerIndex := (ackCount) * (helper.GetConfig().ChildBlockInterval)
+		lastCheckpoint, err := keeper.GetCheckpointByIndex(ctx, headerIndex)
+		if err != nil {
+			return nil, sdk.ErrInternal(sdk.AppendMsgToErr(fmt.Sprintf("could not fetch checkpoint by index %v", headerIndex), err.Error()))
+		}
+		start = lastCheckpoint.EndBlock + 1
+	}
+
+	end := start + helper.GetConfig().AvgCheckpointLength
+	rootHash, err := types.GetHeaders(start, end)
+	if err != nil {
+		return nil, sdk.ErrInternal(sdk.AppendMsgToErr(fmt.Sprintf("could not fetch headers for start:%v end:%v error:%v", start, end, err), err.Error()))
+	}
+	accs := sk.GetAllDividendAccounts(ctx)
+	accRootHash, err := types.GetAccountRootHash(accs)
+	if err != nil {
+		return nil, sdk.ErrInternal(sdk.AppendMsgToErr(fmt.Sprintf("could not get generate account root hash. Error:%v", err), err.Error()))
+	}
+	checkpointMsg := types.NewMsgCheckpointBlock(proposer.Signer, start, start+helper.GetConfig().AvgCheckpointLength, hmTypes.BytesToHeimdallHash(rootHash), hmTypes.BytesToHeimdallHash(accRootHash))
```

Comment:
The proposer should be local address instead of actual proposer otherwise it will at https://github.com/maticnetwork/heimdall/blob/develop/checkpoint/types/msg.go#L55-L57

Or our plan is to allow only the current local address as proposer?

Reply:
Cool, so if I am not the proposer I will revert for now.